### PR TITLE
KAFKA-15451: Send offline dirs in broker heartbeat

### PR DESCRIPTION
**Pull request by soarez**
Offline log directories can be propagated via BrokerHeartbeatRequest.

~~The broker halts if it fails to propagate the failure to the controller
within the log directory failure timeout window.~~ Edit: Dir failure timeout to be addressed in [KAFKA-15649](https://issues.apache.org/jira/browse/KAFKA-15649)

Part of KIP-858.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by soarez** [comment]
This PR is marked as draft because the KIP hasn't yet been accepted but it is ready for a review. 

@rondagostino

**Comment by soarez** [comment]
@rondagostino @pprovenzano can you please take a look at this small change?

File: core/src/main/scala/kafka/server/BrokerLifecycleManager.scala [hunk 1]
```diff
@@ -410,6 +444,7 @@ class BrokerLifecycleManager(
         val errorCode = Errors.forCode(message.data().errorCode())
         if (errorCode == Errors.NONE) {
           failedAttempts = 0
+          eventQueue.append(new OfflineDirsClearEvent(dirsInFlight))
```
Comment (pprovenzano) [inline core/src/main/scala/kafka/server/BrokerLifecycleManager.scala]: Do we always want to queue an event even if dirsInFlight is empty?

**Review by pprovenzano — COMMENTED** [review]
One minor comment, otherwise LGTM

File: core/src/main/scala/kafka/server/BrokerLifecycleManager.scala (hunk 1 above)
Comment (soarez) [reply core/src/main/scala/kafka/server/BrokerLifecycleManager.scala]: Good point. I missed that!

**Comment by soarez** [comment]
I realized the controller request handlers were incorrectly manipulating unsynchronized state which should be exclusive to the event handler thread, so I've just made a change to address that. PTAL

@cmccabe if you could also review this please

**Review by pprovenzano — APPROVED** [review]
LGTM

**Comment by cmccabe** [comment]
I believe this is a duplicate of #14392

**Pull request closed without merging**